### ISSUE TITLE
Add Recoverability capabilities to Gateway Satellite

### ISSUE THREAD

**Issue by MarcinHoppe** [Particular/NServiceBus.Gateway#41]
Taskforce: @boblangley @mikeminutillo @janovesk 

## Problem

Currently satellites, including the gateway, have different recoverability requirements from standard endpoints and cannot share FLR, SLR, and StoreFaultsInErrorQueue behaviors.

The approach taken by the Core is for each satellite to implement an internal recoverability behavior to implement its own specific requirements (e.g. Timeouts only use FLR and move failed messages to error queue).

Another aspect is that we don't want users to be notified about FLR/SLR/Fault events that occur in satellites. 

## The task at hand

In case of the gateway, the requirement is to skip FLR but use SLR and move failed messages to error queue. This rationale behind this approach is the fact that the prevalent failure mode of the gateway is that the channel between sites is down or faulty. In such case immediate message retries would put even more pressure on the faulty channel. Exponential backoff is a much better error handling policy in this situation.

## Plan of attack

- [x] Implement the new recoverability behavior - [PR](https://github.com/Particular/NServiceBus.Gateway/pull/46)
- [x] Creat [core PR](https://github.com/Particular/NServiceBus/pull/3688) to extend FaultContext
  - [x] Created [alternative approach for core PR](https://github.com/Particular/NServiceBus/pull/3694)
- [x] Get [core PR]https://github.com/Particular/NServiceBus/pull/3694) approved and merged
- [x] Finalize the disussions on [Suggested PR](https://github.com/Particular/NServiceBus/pull/3688)
- Document the new config settings 
- [x] Decide whether the behavior needs to be configurable using SLR config API or if we can come up with sensible defaults. Figure out what the defaults for retries/timeIncrease should be if wanted.
- [x] Doco [PR](https://github.com/Particular/docs.particular.net/pull/1436)
 * New behaviour
 * Config API
 * Upgrade guide PR

## Retrospective

Attending: @janovesk @boblangley

### What went well?

* Regular handoffs / updates
* Low barrier to refactor others code

### What can be improved?

* Smaller doco PRs
* Working over 3 continents can cause design churn. Members of Task Forces with 3 continents should be prepared for more drastic changes when it is their turn to contribute.
* No estimate for size was provided, and it felt like more effort was needed than expected.

**Comment by timbussmann** [comment]
As long as the merging of the retry behaviors haven't happen (and even afterwards, depending on the configuration options), why is there exactly the need of a custom retry behavior right now? What retry mechanisms does the gateway currently use on which pipelines?

**Comment by MarcinHoppe** [comment]
@timbussmann @andreasohlund I updated issue description to clearly indicate why we need a recoverability behavior in the gateway.

**Comment by andreasohlund** [comment]
@janovesk did you plan to take a stab at this one?

**Comment by janovesk** [comment]
My unstanding was that @MarcinHoppe was doing this? @andreasohlund

**Comment by andreasohlund** [comment]
He's working on the SLR bug in the core and the sql patch release, @MarcinHoppe ?

**Comment by MarcinHoppe** [comment]
@andreasohlund @janovesk All you say is true. This issue is next on my priority list after the SLR bug in the core. I already teamed up with @SzymonPobiega here on gateway recoverability.

**Comment by andreasohlund** [comment]
@janovesk what's your commitments atm? would you be able to get it going until Marcin can join?

**Comment by boblangley** [comment]
We are including in scope for v6.

/cc: @Particular/tf-v6-launch-plat-dev

**Comment by andreasohlund** [comment]
@Particular/platform-dev-squad this is a beta blocker for the gateway. Propose we move this next in line

**Comment by adamralph** [comment]
@andreasohlund agreed, done.

**Comment by MarcinHoppe** [comment]
I'll be off for a couple of days now so removing myself from the TF.

**Comment by boblangley** [comment]
I am interested in joining this task force.

**Comment by mikeminutillo** [comment]
I have spent some time looking at satellite pipeline stuff recently so I can join too.

**Comment by janovesk** [comment]
I'm good to go!

**Comment by mikeminutillo** [comment]
Created a PR with a start at an implementation. Forwards failures straight to error queue. Here's a breakdown of the issues I have run in to so far:

- [How do we find the configured error queue?](https://github.com/Particular/NServiceBus.Gateway/pull/46/files#diff-446c6e75faefcf9536b67ce699048a2dR65)
- [What is the best way to set the exception headers?](https://github.com/Particular/NServiceBus.Gateway/pull/46/files#diff-4d3f6258fd3de949fccbdb37f0b51fe8R40)
- [We are bypassing the fault host info behavior](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Hosting/FaultHostInformationBehavior.cs). To be fair though, [so does the timeout recoverability behavior](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/DelayedDelivery/TimeoutManager/TimeoutRecoverabilityBehavior.cs#L69).

Next steps after this are to add in some of the [SLR behavior](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Recoverability/SecondLevelRetries/SecondLevelRetriesBehavior.cs#L50) with a few modifications:
- Speaking with @janovesk, we don't believe we need to be concerned with the receive transaction. Basically, if it fails, we should be able to send the message straight to the timeout manager. This is because all of the code inside the receive transaction is HTTP calls to the other side of the gateway which don't participate in a transaction anyway and have their own de-duplication.
- The existing behavior is set up as a fork connector. We can't do that as the part of the pipeline we'd be forking to does not exist in the context of the satellite pipeline. Basically, there is no easy way to skip from an incoming satellite behavior to a main pipeline outgoing behavior. Instead we'll need to route directly to the timeout manager by [doing this](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/DelayedDelivery/TimeoutManagerRoutingStrategy.cs) followed by a direct dispatch.

**Comment by mikeminutillo** [comment]
After a chat with Szymon, it seems we can use a Fork behavior to fall back to the main pipeline it we want. https://github.com/Particular/NServiceBus/pull/3680 for an example of doing that.

**Comment by SzymonPobiega** [comment]
@mikeminutillo this pull request needs to be merged before you can use this because an extension method for creating routing context was not available on satellite level:(

**Comment by SzymonPobiega** [comment]
TF, feel free to cherrypick https://github.com/Particular/NServiceBus/pull/3680 to the `release-6.0.0` branch or, if you don't have bandwidth, tell me to do it.

**Comment by mikeminutillo** [comment]
@SzymonPobiega Why do we limit the source context for the `CreateXContext` methods? Couldn't we make it a generic and cover all cases?

```CSharp
public static IFaultContext CreateFaultContext<TSourceContext>(this ForkConnector<TSourceContext, IFaultContext> forkConnector, TSourceContext sourceContext, OutgoingMessage outgoingMessage, string errorQueueAddress, Exception exception) where TSourceContext : IBehaviorContext
{
    return new FaultContext(outgoingMessage, errorQueueAddress, exception, sourceContext);
}
```

Also, would it make sense for the the error queue address to be set by a behavior in the `IFaultContext`? That isolates the knowledge of where that queue is to a single location. Forking behaviors just need to forward to the Fault Context with details of the error.

**Comment by mikeminutillo** [comment]
The same for Setting the Exception headers really.

**Comment by mikeminutillo** [comment]
Implemented SLR style retries and Forwarding to Error Queue as ForkContext behaviors from the Transport Receive context of the Satellite pipeline. This appears to work. 

TODO:

- Create a PR against core to move some code into Fault Context behaviors. If we do this then forwarding to the Fault Context is good enough.
  - Routing Faulted messages to the Error Queue. This would prevent [having to know where the error queue is](https://github.com/Particular/NServiceBus.Gateway/pull/46/files#diff-446c6e75faefcf9536b67ce699048a2dR65). This is currently hard-coded to "error".
  - Setting Faulted message headers. This would prevent [having to expose the current internal implementation](https://github.com/Particular/NServiceBus.Gateway/pull/46/files#diff-4d3f6258fd3de949fccbdb37f0b51fe8R40). These are currently skipped.
- Write some acceptance tests around faulted gateway connections. Presumably just opening one end and forwarding to sites that aren't running is good enough for this. 

NOTES:
- We don't need to cherry pick https://github.com/Particular/NServiceBus/pull/3680 to `release-6.0.0` because we are forking from a Transport Receive Context instead of from a Satellite Processing Context (as the PR is doing).
- We are no longer bypassing the Fault Context so the behavior that sets faulted info is fine
- In discussion with @danielmarbach and @andreasohlund we decided that we should move the Fault info behaviors in the Fault Context as listed above. We also discussed using SLR headers and decided to do it as they represent the same thing. 

I have set up the retries to occur over 4 minutes (total). Is this good enough? Will most Gateway link outages recover in this time or does it need to be longer? Do we want this to be more configurable?

**Comment by mikeminutillo** [comment]
Also note, I have set up the Gateway to depend on the "DelayedDelivery" feature. This feature can be enabled even when it won't actually work which is interesting. This is the case where the TimeoutManager is disabled and the transport does not handle delayed delivery natively. In such cases, the [`ThrowIfCannotDeferMessageBehavior`](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/DelayedDelivery/ThrowIfCannotDeferMessageBehavior.cs) gets installed into the Routing Context. As we are forwarding to the Routing Context, we will hit that. I'm just mentioning it as it means that for MSMQ (where you are likely testing), you will need to enable the Timeout Manager to get the delayed retries to work. This should be mentioned in the doco for the Gateway.

**Comment by janovesk** [comment]
Did you see [this comment](https://github.com/Particular/NServiceBus/pull/3680#issuecomment-210759976) about skipping pipeline completely @mikeminutillo ?

**Comment by andreasohlund** [comment]
@janovesk : already talked to mike about it and we're good (I still thing we can get rid of the pipe in the future but that would mean pushing retries and error queue movement into the transports)

I'll raise new issues to discuss this as needed

In short: I'm good with the current approach

**Comment by janovesk** [comment]
Update:
* [Suggested PR](https://github.com/Particular/NServiceBus/pull/3688) to the core about extending FaultContext to avoid leaking the error queue outside the core.
* Updated this branch to use that PR.
* Added config API for retries. `GatewaySettings.Retries`
* Added test that checks that messages are moved to error queue when they fail with the new error handling.

TODO:

- Add test that checks the configured number of retries happens before moving to error queue
- Finalize the discussions on [Suggested PR](https://github.com/Particular/NServiceBus/pull/3688)
- Document the new config settings 
- Discuss what the defaults for retries/timeIncrease should be.

**Comment by mikeminutillo** [comment]
Update:
  - Discussed alternative approach about extending FaultContext with the core maintainers and implemented it [here](https://github.com/Particular/NServiceBus/pull/3694)

TODO:
  - Select one of the Core PRs to go with and then as above.

**Comment by janovesk** [comment]
Update:
* Closed initally proposed [Suggested PR](https://github.com/Particular/NServiceBus/pull/3688). The [new PR](https://github.com/Particular/NServiceBus/pull/3694) by @mikeminutillo is an obvious better option.
* Updated this branch to use the [new PR](https://github.com/Particular/NServiceBus/pull/3694).
* Added test that checks the configured number of retries happens before moving to error queue
* Fixed a lurking bug in the gateway where routing based on originating site header was applied to any message put on the gw satellite. This would lead to infinite loop as we route to ourselves.

TODO:

- Sit on the core maintainers until they approve the [new PR](https://github.com/Particular/NServiceBus/pull/3694)
- Document the new config settings 
- Discuss what the defaults for retries/timeIncrease should be.

**Comment by boblangley** [comment]
Update:
* The [new PR](https://github.com/Particular/NServiceBus/pull/3694) has not yet been merged.
* I added commits to allow implementation of a custom retry policy and to make disabling retries easier to discover.
* Doco PR for the new FaultContext PR has been created: https://github.com/Particular/docs.particular.net/pull/1435
* Doco PR for Gateway recoverability configuration has been created: https://github.com/Particular/docs.particular.net/pull/1436

TODO:

- Sit on the core maintainers until they approve the [new PR](https://github.com/Particular/NServiceBus/pull/3694)
- Update Gateway with core NuGets.
- Update FaultContext doco PR with core Nugets and have doco maintainers merge.
- Finish documenting the new config settings (including Gateway nugets updated for snippets)
- Discuss what the defaults for retries/timeIncrease should be.

**Comment by janovesk** [comment]
Update:
Made some changes to the [doco pr](https://github.com/Particular/docs.particular.net/pull/1436). Added notice in the upgrade guide.

TODO:
Update Gateway with core NuGets.
Update FaultContext doco PR with core Nugets and have doco maintainers merge.
Finish documenting the new config settings (including Gateway nugets updated for snippets)

**Comment by janovesk** [comment]
Core PR has been merged now @boblangley. Can you upgrade the gateway with the new package?

**Comment by janovesk** [comment]
As to "Discuss what the defaults for retries/timeIncrease should be". I think the backof strategy suggested by @boblangley with adding an additonal 60 seconds for each retry is good. 

We also provide the flexibility of plugging in a custom retry policy, enabling customers to change it as they see fit.

**Comment by mikeminutillo** [comment]
I'm happy with the back-off strategy. This means we'll retry for up to 5 minutes. 

The Gateway is up to date. I'm just not sure about having the Gateway expose an extension method to get retry count. I will push up a change to pass in the retry count to the recoverability policy instead and then we can get this merged and close out the docs PR. 

I believe the core maintainers are responsible for this repo. I'll push up my proposed changes and ping them.

**Comment by mikeminutillo** [comment]
Modified the issue name to better reflect what we are doing rather than our desired end state

**Comment by boblangley** [comment]
Update:

Made some changes to the PR to address maintainer concerns.

Todo:

Address remaining concerns.

**Comment by boblangley** [comment]
Update:

Addressed maintainer comments or asked for clarification.
Added another commit to the doco update to address most of maintainer remarks. Left open question about a link (not sure which one to use).
Closed FaultContext PR until Beta2 nuget is available.

Todo:

Get Gateway merged.
Get Gateway doco completed / merged.

Beta Nuget:

This Beta will require Beta2 nugets. @andreasohlund @danielmarbach how do we proceed?

**Comment by andreasohlund** [comment]
>This Beta will require Beta2 nugets. @andreasohlund @danielmarbach how do we proceed?

@Particular/nservicebus-maintainers let's release a beta 2 from the current release-6.0.0 branch?

@boblangley is it needed asap or is this a blocker for this task?

**Comment by andreasohlund** [comment]
Beta2 is now ready on the buildserver, https://builds.particular.net/viewLog.html?buildId=161217&buildTypeId=NServiceBusCore_Build&tab=artifacts#!12q45qa , please smoke test it and give me the green light to upload it to myget

**Comment by boblangley** [comment]
@andreasohlund I smoke tested and it looks good.

**Comment by SimonCropp** [comment]
doco merged. should this now be closed?

**Comment by janovesk** [comment]
We'll do a quick retrospective and close this @SimonCropp.

**Comment by boblangley** [comment]
@janovesk and I did a retrospective. The notes are added to the description.
